packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -405,6 +430,108 @@ contract ControllerHelper is FlashControllerHelper, IERC721Receiver {
+                params.wPowerPerpAmount,
+                params.collateralToMint,
+                params.collateralToLp,
+                params.collateralToWithdraw,
+                params.amount0Min,
+                params.amount1Min,
+                uint256(params.lowerTick),
+                uint256(params.upperTick)
+            )
+        );
+
+        console.log("IWETH9(weth).balanceOf(address(this) when exit", IWETH9(weth).balanceOf(address(this)));
+    }
+
+    function _flashCallback(
+        address, /*_initiator*/
+        address _asset,
+        uint256 _amount,
+        uint256 _premium,
+        uint8 _callSource,
+        bytes memory _calldata
+    ) internal override {
+        if (CALLBACK_SOURCE(_callSource) == CALLBACK_SOURCE.FLASHLOAN_W_MINT_DEPOSIT_NFT) {
+            console.log("_amount", _amount);
+            console.log("_premium", _premium);
+            console.log(
+                "IWETH9(weth).balanceOf(address(this) when receiving flashloan",
+                IWETH9(weth).balanceOf(address(this))
+            );
+
+            FlashloanMintDepositNftData memory data = abi.decode(_calldata, (FlashloanMintDepositNftData));
+
+            // convert flashloaned WETH to ETH
+            IWETH9(weth).withdraw(_amount);
+
+            uint256 vaultId = IController(controller).mintWPowerPerpAmount{value: data.collateralToMint}(
+                data.vaultId,
+                data.wPowerPerpAmount,
+                0
+            );
+
+            // LP data.wPowerPerpAmount & data.collateralToLp in Uni v3
+            uint256 amount0Desired = isWethToken0 ? data.collateralToLp : data.wPowerPerpAmount;
+            uint256 amount1Desired = isWethToken0 ? data.wPowerPerpAmount : data.collateralToLp;
+            uint256 uniTokenId = _lpWPowerPerpPool(
+                address(this),
+                data.collateralToLp,
+                amount0Desired,
+                amount1Desired,
+                data.lpAmount0Min,
+                data.lpAmount1Min,
+                block.timestamp,
+                int24(data.lpLowerTick),
+                int24(data.lpUpperTick)
+            );
+            console.log("uniTokenId", uniTokenId);
+
+            // deposit Uni NFT token in vault
+            INonfungiblePositionManager(nonfungiblePositionManager).approve(controller, uniTokenId);
+            IController(controller).mintWPowerPerpAmount(vaultId, 0, uniTokenId);
+
+            console.log("address(this).balance just before withdrawing ETH", address(this).balance);
+            console.log(
+                "IWETH9(weth).balanceOf(address(this) just before withdrawing ETH",
+                IWETH9(weth).balanceOf(address(this))
+            );
+
+            // remove flashloan amount in ETH from vault + any amount of collateral user want to withdraw (sum <= vault.collateralAmount)
+            IController(controller).burnWPowerPerpAmount(vaultId, 0, _amount.add(data.collateralToWithdraw));
```

Comment:
could use withdraw() cheaper?

I think there are a couple use cases that we can chat on tmrw, but think that in most cases a user will want to deposit some ETH (say less than 150% CR without the NFT, so a flashloan neede still) in their vault and a more edge case is withdrawing collateral (as per the above, which would be if they had a vault that already had some collateral in it and wanted to LP).

I also think that a user shouldn't necessarily need to deposit msg.value > params.collateralToLP as its possible a user uses the collateralToWithdraw for their LP (and can effectively use a flashloan for this as well that is repaid later when withdrawn from vault)